src/krylov/pcg.c

```diff
@@ -565,35 +588,53 @@ hypre_PCGSolve( void *pcg_vdata,
       /* alpha should always be greater zero for spd A, spd precond. and nonzero p, r*/
       /* alpha = <r, Cr> / <p, Ap> */
       /* for alpha close to 0 x and r will not change much unless x and/or r are close to 0 */
-      if (! (alpha >= HYPRE_REAL_MIN) )
+      if (alpha <=0.0)
+      {
+         if (print_level > 1 && my_id == 0)
+	 {
+            hypre_printf("alpha %e", alpha);
+	 }
+         hypre_error_w_msg(HYPRE_ERROR_CONV, "Negative or zero alpha value in PCG");
+         if (skip_break == 0)
+	 {
+	    if (i == 1) { i_prod = i_prod_0; }
+            break;
+	 }
+      }
+      else if (! (alpha  >= HYPRE_REAL_TRUE_MIN))
+      {
+         hypre_error_w_msg(HYPRE_ERROR_CONV, "alpha value less than TRUE_MIN in PCG");
+         if (print_level > 1 && my_id == 0)
+	 {
+            hypre_printf("alpha %e", alpha);
+	 }
+         if (skip_break < 2)
+	 {
+	    if (i == 1) { i_prod = i_prod_0; }
+            break;
+	 }
+      }
+      else if (! (alpha  >= HYPRE_REAL_MIN))
       {
          hypre_error_w_msg(HYPRE_ERROR_CONV, "Subnormal alpha value in PCG");
-         if (i == 1) { i_prod = i_prod_0; }
-         if (!skip_break)
-         {
+         if (print_level > 1 && my_id == 0)
+	 {
+            hypre_printf("alpha %e", alpha);
+	 }
+         if (skip_break < 3)
+	 {
+	    if (i == 1) { i_prod = i_prod_0; }
             break;
-         }
-         else if (! (alpha  >= HYPRE_TRUE_MIN))
-         {
-            hypre_error_w_msg(HYPRE_ERROR_CONV, "alpha value less than TRUE_MIN in PCG");
-            if (skip_break == 1) break;
-         }
-         else if (alpha <=0.0)
-         {
-            if (print_level > 1 && my_id == 0)
-	    {
-               hypre_printf("alpha %e", alpha);
-	    }
-            hypre_error_w_msg(HYPRE_ERROR_CONV, "Negative or zero alpha value in PCG");
-            if (skip_break == 2) break;
-         }
+	 }
       }
 
       gamma_old = gamma;
 
       /* x = x + alpha*p */
       (*(pcg_functions->Axpy))(alpha, p, x);
 
+      if (flex) (*(pcg_functions->CopyVector))(r, r_old); /*save old residual */
```

Comment:
You probably should run this through 'astyle'.  I think one-line 'if' statements are okay, but they might require the squirly braces.  I'd have to check again.